src/lib/core/selection/selection.ts

```diff
@@ -0,0 +1,168 @@
+import {Subject} from 'rxjs/Subject';
+
+
+/**
+ * Class to be used to power selecting one or more options from a list.
+ * @docs-private
+ */
+export class MdSelectionModel {
+  constructor(
+    private _options: any[],
+    private _isMulti = false,
+    initiallySelectedValues?: any[]) {
+
+    if (initiallySelectedValues && initiallySelectedValues.length) {
+      if (_isMulti) {
+        initiallySelectedValues.forEach(value => this._select(value));
+      } else {
+        this._select(initiallySelectedValues[0]);
+      }
+
+      // Clear the array in order to avoid firing the change event for preselected values.
+      this._unflushedSelectedValues.length = 0;
+    }
+  }
+
+  /** Event emitted when the value has changed. */
+  onChange: Subject<MdSelectionChange> = new Subject();
+
+  /** Currently-selected values. */
+  private _selectedValues: any[] = [];
```

Comment:
Rather than keeping an array of `selectedValues`, we should keep the selection in a `Set` (named `_selection`). We can use `.keys()` when we want the list of selected values (caching the value). This will make selecting, deselecting, and checking selection much faster (and the code simpler). When the options are updated, the update would then look like.

Overall this would change insertion, deletion from O(k) to O(1) and keep getting an array of the selected values mostly the same.